gcp_variant_transforms/transforms/extract_variant_header.py

```diff
@@ -0,0 +1,137 @@
+
+
+__all__ = ['ExtractHeaderFieldsFromVariant']
+
+
+_FLAG = 'Flag'
+_STRING = 'String'
+_NO_DESCRIPTION = 'no_description'
+_UNKNOWN_SOURCE = 'unknown_source'
+_UNKNOWN_VERSION = 'unknown_version'
+
+
+class _ExtractHeaderFieldsFromVariant(beam.DoFn):
+  """Extracts header fields from ``Variant`` record."""
+
+  def _is_defined_info_field(self, field_key, headers):
+    return headers and field_key in headers.infos
+
+
+  def _is_defined_format_field(self, field_key, headers):
+    return headers and field_key in headers.formats
+
+
+  def _get_field_counts(self, field_value):
+    # TODO(nmousavi): return filed count by conditioning on field_value type
+    # (list, bool, o/w) once variant_to_bigquery code can deduct type for
+    # missing header fields from extracted final headers. Currently it inserts
+    # fields with missing header to bigquery as a repeated field.
+    return field_counts['.']
+
+
+  def _get_field_type(self, field_value):
+    if isinstance(field_value, bool):
+      return _FLAG
+    else:
+      return _STRING
+
+
+  def _extract_info_fields(self, variant, headers):
+    infos = {}
+    for info_field_key, variant_info in variant.info.iteritems():
+      info_field_value = variant_info.data
+      if not self._is_defined_info_field(info_field_key, headers):
+        assert info_field_key not in infos, (
+            "Bad VCF file. Duplicate INFO field. %s" %info_field_key)
+        infos[info_field_key] = Info(info_field_key,
+                                     self._get_field_counts(info_field_value),
+                                     self._get_field_type(info_field_value),
+                                     _NO_DESCRIPTION,
+                                     _UNKNOWN_SOURCE,
+                                     _UNKNOWN_VERSION)
+    return infos
+
+
+  def _extract_format_fields(self, variant, headers):
+    formats = {}
+    for call in variant.calls:
+      for format_key, format_value in call.info.iteritems():
+        if not self._is_defined_format_field(format_key, headers):
+          assert format_key not in formats, (
+              "Bad VCF file. Duplicate FORMAT field. %s" %format_key)
+          formats[format_key] = Format(format_key,
+                                       self._get_field_counts(format_value),
+                                       self._get_field_type(format_value),
+                                       _NO_DESCRIPTION)
+      # No point in proceeding. All other calls have the same FORMAT.
+      break
+    return formats
+
+
+  def _extract_header_fields(self, variant, headers):
+    infos = self._extract_info_fields(variant, headers)
+    formats = self._extract_format_fields(variant, headers)
+    if infos or formats:
+      return [vcf_header_io.VcfHeader(infos=infos, formats=formats)]
+    else:
+      return []
+
+
+  def process(self, variant, headers):
+    return self._extract_header_fields(variant, headers)
```

Comment:
I think the body of self._extract_header_fields is simple enough to be inlined here. That way, you can use `yield` instead of sometimes returning empty list as well.

Reply:
Done